### PR TITLE
Add fully-derived request readers for case classes

### PR DESCRIPTION
**Pull request by travisbrown**
This isn't a candidate for merging at the moment, just a demo for tonight's meetup that we may want to consider including someday (any feedback is welcome, of course).

### CODE REVIEW

File: core/src/main/scala/io/finch/request/RequestReader.scala [hunk 1]
```diff
@@ -205,4 +205,36 @@ object RequestReader {
...
+      ct: ClassTag[HV],
+      key: Witness.Aux[HK],
+      fpt: FromParams[T]
+    ): FromParams[FieldType[HK, HV] :: T] = new FromParams[FieldType[HK, HV] :: T] {
+      def reader: RequestReader[FieldType[HK, HV] :: T] =
+        param(key.value.name).as(dh, ct).map(field[HK](_)) :: fpt.reader
+    }
+  }
+
+  def to[A]: GenericDerivation[A] = new GenericDerivation[A]
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/request/RequestReader.scala]: Don't you think `for[User]` feels more suitable here? I remember it was `for` in the discussion but I'm not sure which name is better.

File: core/src/main/scala/io/finch/request/RequestReader.scala [hunk 2]
```diff
@@ -205,4 +205,36 @@ object RequestReader {
...
+      gen: LabelledGeneric.Aux[A, Repr],
+      fp: FromParams[Repr]
+    ): RequestReader[A] = fp.reader.map(gen.from)
+  }
+
+  trait FromParams[L <: HList] {
+    def reader: RequestReader[L]
+  }
+
+  object FromParams {
```
Comment (vkostyukov) [inline core/src/main/scala/io/finch/request/RequestReader.scala]: I like it! Do you think we should do the same trick for headers (is it useful?) and bodies. Like:

```scala
val a = RequestReader.for[User].fromParams
val b = RequestReader.for[User].fromHeaders
val c = RequestReader.for[User].fromBody
```

File: core/src/main/scala/io/finch/request/RequestReader.scala (hunk 1 above)
Comment (travisbrown) [inline core/src/main/scala/io/finch/request/RequestReader.scala]: Yeah, but `for` is a keyword, which I'd forgotten during that conversation, apparently. :smile:

**Comment by vkostyukov** [comment]
This is amazing! Thanks, Travis!

I like the API really much. And I think if we extend it to headers and bodies it will be very intuitive.

File: core/src/main/scala/io/finch/request/RequestReader.scala (hunk 1 above)
Comment (vkostyukov) [inline core/src/main/scala/io/finch/request/RequestReader.scala]: Oh my :) I completely forgot about it :) I should use it more to remember!

Comment (vkostyukov) [inline core/src/main/scala/io/finch/request/RequestReader.scala]: How about `RequestReader.of[User]`?

Comment (vkostyukov) [inline core/src/main/scala/io/finch/request/RequestReader.scala]: Or even: 

`RequestReader.materialize[User].fromParams`
`RequestReader.derive[User].fromParams`

**Comment by travisbrown** [comment]
Now included in #266.

**Pull request closed without merging**